spec/support/test_data_cache.rb

```diff
@@ -0,0 +1,55 @@
+# rubocop:disable Style/ClassVars
+class TestDataCache
+  #  A helper class for pre-creating test data with FactoryBot factories
+  #  and maintaining a cache of these objects to avoid the performance problems
+  #  of repeatedly asking FactoryBot to create trees of objects.
+  #
+  #  Refs: https://evilmartians.com/chronicles/testprof-2-factory-therapy-for-your-ruby-tests-rspec-minitest
+  #
+  # The test data and can be accessed as follows:
+  #
+  # TestSetupCache.get(:course, :primary, :unpublished)
+  #
+  # Note that a course with the requested traits needs to be pre-defined in this class.
+  # It does not matter what order the traits are when passed as arguments.
+
+
+  class << self
+    def get(name, *traits)
+      case name
+      when :course
+        if @@courses == nil
+          # cache not populated, fall back to factory (while we migrate all our tests)
```

Comment:
Would it make more sense to test this on a factory-by-factory basis? Something like the following, replacing the `if @@courses == nil` check above.

```ruby
          sorted_traits = traits.sort
          unless @@courses.has_key? sorted_traits
            if @@course_factories.has_key? sorted_traits
              @@courses[sorted_traits] = @@course_factories[sorted_traits].call
            else
              raise_trait_error(:course, traits)
            end
          end
          @@courses[sorted_traits]
```

If you agree, it's probably worth putting most of the above in a separate function.

Reply:
I'm don't think this is necessary ...
Unless the caching is switched on then `@@courses` will always be nil. Therefore it probably makes sense to simply return the factory.
What do you think @timabell?

Reply:
What @TobyRet said.

Because the behaviour is:

* The cache can be turned off/on with the env
* If it's on, then the cache is filled in advance so no need to do granular testing
* If it's off, we bypass the cache entirely and fall back to the factories, so again no need to test what is cached

Having said that, this isn't immediately obvious from the code as written, I think maybe we should add a `cachePopulated` field to the class and set that in the setup method and then update this `if` to use it for clarity

Reply:
... and / or we can check if the cache is actually turned on? `ENV['TEST_DATA_CACHE']`

Reply:
As I mentioned on the phone, that make the behaviour unnecessarily affected by external state.

I'm having a go at implementing this but my first attempt broke something...

Reply:
Refactored, all green. Should be clearer now.

Reply:
Looks good to me.